Features/#148 multiline function call in a single-line declared section

dakochik opened the pull request:
Issue #148 is resolved

iromeo commented on src/main/kotlin/com/jetbrains/snakecharm/inspections/SmkMultilineFunctionCallInspection.kt:
```diff
@@ -0,0 +1,76 @@
+                return
+            }
+
+            val problemCalls = arrayListOf<PsiElement>()
```
here whitespaces are collected, not PyCallExpressions, so better to rename variable into `invalidWhitespaces`

iromeo commented on src/main/kotlin/com/jetbrains/snakecharm/inspections/SmkMultilineFunctionCallInspection.kt:
```diff
@@ -0,0 +1,76 @@
+            val problemCalls = arrayListOf<PsiElement>()
+            args.children.filter { it.elementType == PyElementTypes.CALL_EXPRESSION }.forEach {
+                multilineFunctionCall(it.node, problemCalls)
+            }
```
let's reduce low level PSI access, e.g. you could filter PyCallExpressions and use API to get arg list:
```kotlin
            args.arguments.forEach { psi ->
                if (psi is PyCallExpression) {
                    collectNewlinesInMultilineCall(psi, invalidWhitespaces) // collectNewlinesInMultilineCall(PyCallExpression, MutableList<PsiElement>)
                }
            }
```

iromeo commented on src/main/kotlin/com/jetbrains/snakecharm/inspections/SmkMultilineFunctionCallInspection.kt:
```diff
@@ -0,0 +1,76 @@
+        override fun getFamilyName() = SnakemakeBundle.message("INSP.NAME.multiline.func.call.fix")
+
+        override fun applyFix(project: Project, descriptor: ProblemDescriptor) {
+            WriteCommandAction.runWriteCommandAction(project) { descriptor.psiElement.delete() }
```
Doesn't work for case:
```
rule rule_148_b:
    input: foo1("text1", #ff
                "text2", "text3")
```

iromeo commented on src/main/kotlin/com/jetbrains/snakecharm/inspections/SmkMultilineFunctionCallInspection.kt:
```diff
@@ -0,0 +1,76 @@
+            problemCalls.forEach {
+                registerProblem(
+                    it,
+                    SnakemakeBundle.message("INSP.NAME.multiline.func.call"),
```
For:
```
rule rule_148_b:
    input: foo1("text1",
                "text2", "text3")
```
2 valid froms available:
```
rule rule_148_b:
    input: foo1("text1","text2","text3")
```
and
```
rule rule_148_b:
    input: 
        foo1("text1",
                "text2", "text3")
```
At the moment the error message forces single line and it could be confusing, if 2nd variant was assumed

iromeo reviewed (changes requested):
Not everything works as expected

dakochik commented:
fixed: quickfix was modified, so now it suggests to transform single-line declared section to multiline declared as follows:
```
# before:
rule rule_148_b:
    input:  foo1("text1",
                "text2", "text3"), foo2("text4")

# after:
rule rule_148_b:
    input: 
        foo1("text1",
            "text2", "text3"),
        foo2("text4")
```

refactored: `multilineSectionDefinition()` was moved into `SmkRuleOrCheckpointArgsSection`, see commit message, some minor changes in inspection code, which were requiered

iromeo commented on src/main/kotlin/com/jetbrains/snakecharm/lang/psi/impl/SmkRuleOrCheckpointArgsSectionImpl.kt:
```diff
@@ -57,4 +57,24 @@ open class SmkRuleOrCheckpointArgsSectionImpl(node: ASTNode): SmkArgsSectionImpl
+            while (true) {
+                node = when (node.elementType) {
+                    TokenType.WHITE_SPACE -> {
+                        if (node.text.contains("\n")) {
```
better to use char instead of single char string `node.text.contains('\n')`

iromeo commented on src/main/kotlin/com/jetbrains/snakecharm/inspections/SmkMultilineFunctionCallInspection.kt:
```diff
@@ -0,0 +1,98 @@
+import com.jetbrains.snakecharm.SnakemakeBundle
+import com.jetbrains.snakecharm.lang.psi.SmkRuleOrCheckpointArgsSection
+
+const val TAB = "    "
```
We cannot use such TAB def here, number of spaces in TAB is IDE setting. More over user could use real TAB character instead of spaces.

iromeo replied:
And is only setting for format code features, indents in user file could be different
![image](https://user-images.githubusercontent.com/72933/127317233-a91e8f6a-75b8-4bbd-98a1-601a5b5850fa.png)

iromeo commented on src/main/kotlin/com/jetbrains/snakecharm/inspections/SmkMultilineFunctionCallInspection.kt:
```diff
@@ -0,0 +1,98 @@
+            val indent = startElement.prevSibling.text
+            // Moves every argument list element to new line
+            argumentList.arguments.forEach { expression ->
+                doc.insertString(expression.startOffset, "$indent$TAB")
```
I think is better to use `indent` instead of TAB here, because actually you need to indent other lines relative to section name statement

iromeo reviewed (changes requested):
Please get rid of harcoded TAB meaning. In general everything works good.

dakochik commented:
fixed: TAB hardcoding, 'contains' function calling
